### PR TITLE
Fix crash in lbmap due to backendsByMapIndex containing holes

### PR DESCRIPTION
**Pull request by tgraf**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6558)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: pkg/maps/lbmap/bpfservice.go [hunk 1]
```diff
@@ -124,14 +134,24 @@ func (b *bpfService) deleteBackend(backend ServiceValue) {
...
-		backends[i-1] = b.backendsByMapIndex[i].bpfValue
+		if b.backendsByMapIndex[i] == nil {
+			log.Errorf("BUG: hole found in backendsByMapIndex: %#v", b.backendsByMapIndex)
+			continue
+		}
+
+		backends[dstIndex] = b.backendsByMapIndex[i].bpfValue
+		dstIndex++
 	}
+	b.mutex.RUnlock()
```
Comment (vadorovsky) [inline pkg/maps/lbmap/bpfservice.go:128]: `defer b.mutex.RUnlock()`, just for consistency?

File: pkg/maps/lbmap/bpfservice.go [hunk 2]
```diff
@@ -215,5 +241,7 @@ func (l *lbmapCache) prepareUpdate(fe ServiceKey, backends []ServiceValue) *bpfS
 }
 
 func (l *lbmapCache) delete(fe ServiceKey) {
+	l.mutex.Lock()
 	delete(l.entries, fe.String())
+	l.mutex.Unlock()
```
Comment (vadorovsky) [inline pkg/maps/lbmap/bpfservice.go:222]: same here

**Review by vadorovsky — APPROVED** [review]

File: pkg/maps/lbmap/bpfservice.go (hunk 1 above)
Comment (tgraf) [reply pkg/maps/lbmap/bpfservice.go:128]: It is faster without the defer and this is the most frequently called function

File: pkg/maps/lbmap/bpfservice.go (hunk 2 above)
Comment (tgraf) [reply pkg/maps/lbmap/bpfservice.go:222]: For single exit functions, there is not much of a point to pay the performance penalty of the defer

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
@aanm identified another issue which calls for a bigger fix that will likely remove this code.

**Comment by tgraf** [comment]
test-me-please

**Review by aanm — APPROVED** [review]
I've looked through all code, the mutex is not needed as both structures are indeed being protected by the caller.

**Pull request merged**